### PR TITLE
Add --ephemeral option

### PR DESCRIPTION
**Pull request by DaanDeMeyer**
Identical to systemd-nspawn's --ephemeral option (and that's what how
we implement this for mkosi boot) but also implemented for qemu.

Also includes  a small commit to clean up `run_shell`.

### CODE REVIEW

**Comment by behrmann** [comment]
Would maybe a [qemu-snapshot](https://wiki.qemu.org/Documentation/CreateSnapshot#Create_a_snapshot) be an alternative way to implement this?

**Comment by DaanDeMeyer** [comment]
It might be but the documentation is extremely sparse. I think we're better off being consistent here and doing the same as nspawn does internally.

File: mkosi.md [hunk 1]
```diff
@@ -734,6 +734,11 @@ details see the table below.
   in systemd-nspawn's man page:
   https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#-n
 
+`--ephemeral`
+: When used with the shell, boot or qemu verbs, this option runs the specified
+  verb on a temporary snapshot of the output image that is removed immediately
+  when the container terminates.
```
Comment (behrmann) [inline mkosi.md:740]: Maybe add here, that it uses the `--ephemeral` switch for nspawn and a copy of the image for qemu

File: mkosi/__init__.py [hunk 2]
```diff
@@ -5947,11 +5956,19 @@ def run_qemu(args: CommandLineArguments) -> None:
...
+            fname = args.output
+
+        cmdline += ["-drive", f"format={'qcow2' if args.qcow2 else 'raw'},file={fname},if=virtio"]
+        cmdline += args.cmdline
 
-    print_running_cmd(cmdline)
+        print_running_cmd(cmdline)
 
-    run(cmdline, execvp=True)
+        run(cmdline, execvp=True)
```
Comment (behrmann) [inline mkosi/__init__.py:5971]: I haven't tested this, but does this clean up properly? From the top of my head I don't know whether the ExitStack can be / will be worked through if we exec here.

Comment (DaanDeMeyer) [reply mkosi/__init__.py:5971]: Shit, I didn't consider that. Maybe we should get rid of the execvp?

Comment (DaanDeMeyer) [reply mkosi/__init__.py:5971]: With a very quick test, this doesn't seem to cause any major problems.

Comment (behrmann) [reply mkosi/__init__.py:5971]: That could be done, I guess. Although I don't know whether there was any other intent behind execing here, besides "mkosi is done and doesn't need to stay around".

Comment (DaanDeMeyer) [reply mkosi/__init__.py:5971]: It makes signal handling somewhat more awkward as there's two processes running now. I'll need to verify this doesn't break CTRL+C when using the qemu gui.

Comment (DaanDeMeyer) [reply mkosi/__init__.py:5971]: This works fine, we already handle the two process scenario in run() correctly. I think we can go ahead and change this. I'll do the same change for mkosi boot

File: mkosi.md (hunk 1 above)
Comment (DaanDeMeyer) [reply mkosi.md]: I wouldn't go into specifics, I can copy paste some more info from the nspawn man page though

Comment (DaanDeMeyer) [reply mkosi.md]: Let's not add too many implementation details. I extended with some more info from nspawn's man page that informs the user that the copy is more efficient on filesystems that support reflink.

**Pull request merged**